File: java/flight/flight-sql/src/main/java/org/apache/arrow/flight/sql/FlightSqlUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.arrow.flight.sql;

import java.util.List;

import org.apache.arrow.flight.ActionType;
import org.apache.arrow.flight.CallStatus;

import com.google.common.collect.ImmutableList;
import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;

/**
 * Utilities to work with Flight SQL semantics.
 */
public final class FlightSqlUtils {

  public static final ActionType FLIGHT_SQL_BEGIN_SAVEPOINT =
      new ActionType("BeginSavepoint",
          "Create a new savepoint.\n" +
              "Request Message: ActionBeginSavepointRequest\n" +
              "Response Message: ActionBeginSavepointResult");

  public static final ActionType FLIGHT_SQL_BEGIN_TRANSACTION =
      new ActionType("BeginTransaction",
          "Start a new transaction.\n" +
              "Request Message: ActionBeginTransactionRequest\n" +
              "Response Message: ActionBeginTransactionResult");
  public static final ActionType FLIGHT_SQL_CREATE_PREPARED_STATEMENT = new ActionType("CreatePreparedStatement",
      "Creates a reusable prepared statement resource on the server. \n" +
          "Request Message: ActionCreatePreparedStatementRequest\n" +
          "Response Message: ActionCreatePreparedStatementResult");

  public static final ActionType FLIGHT_SQL_CLOSE_PREPARED_STATEMENT = new ActionType("ClosePreparedStatement",
      "Closes a reusable prepared statement resource on the server. \n" +
          "Request Message: ActionClosePreparedStatementRequest\n" +
          "Response Message: N/A");

  public static final ActionType FLIGHT_SQL_CREATE_PREPARED_SUBSTRAIT_PLAN =
      new ActionType("CreatePreparedSubstraitPlan",
          "Creates a reusable prepared statement resource on the server.\n" +
              "Request Message: ActionCreatePreparedSubstraitPlanRequest\n" +
              "Response Message: ActionCreatePreparedStatementResult");

  public static final ActionType FLIGHT_SQL_CANCEL_QUERY =
      new ActionType("CancelQuery",
          "Explicitly cancel a running query.\n" +
              "Request Message: ActionCancelQueryRequest\n" +
              "Response Message: ActionCancelQueryResult");

  public static final ActionType FLIGHT_SQL_END_SAVEPOINT =
      new ActionType("EndSavepoint",
          "End a savepoint.\n" +
              "Request Message: ActionEndSavepointRequest\n" +
              "Response Message: N/A");
  public static final ActionType FLIGHT_SQL_END_TRANSACTION =
      new ActionType("EndTransaction",
          "End a transaction.\n" +
              "Request Message: ActionEndTransactionRequest\n" +
              "Response Message: N/A");

  public static final List<ActionType> FLIGHT_SQL_ACTIONS = ImmutableList.of(
      FLIGHT_SQL_CREATE_PREPARED_STATEMENT,
      FLIGHT_SQL_CLOSE_PREPARED_STATEMENT
  );

  /**
   * Helper to parse {@link com.google.protobuf.Any} objects to the specific protobuf object.
   *
   * @param source the raw bytes source value.
   * @return the materialized protobuf object.
   */
  public static Any parseOrThrow(byte[] source) {
    try {
      return Any.parseFrom(source);
    } catch (final InvalidProtocolBufferException e) {
      throw CallStatus.INVALID_ARGUMENT
          .withDescription("Received invalid message from remote.")
          .withCause(e)
          .toRuntimeException();
    }
  }

  /**
   * Helper to unpack {@link com.google.protobuf.Any} objects to the specific protobuf object.
   *
   * @param source the parsed Source value.
   * @param as     the class to unpack as.
   * @param <T>    the class to unpack as.
   * @return the materialized protobuf object.
   */
  public static <T extends Message> T unpackOrThrow(Any source, Class<T> as) {
    try {
      return source.unpack(as);
    } catch (final InvalidProtocolBufferException e) {
      throw CallStatus.INVALID_ARGUMENT
          .withDescription("Provided message cannot be unpacked as " + as.getName() + ": " + e)
          .withCause(e)
          .toRuntimeException();
    }
  }

  /**
   * Helper to parse and unpack {@link com.google.protobuf.Any} objects to the specific protobuf object.
   *
   * @param source the raw bytes source value.
   * @param as     the class to unpack as.
   * @param <T>    the class to unpack as.
   * @return the materialized protobuf object.
   */
  public static <T extends Message> T unpackAndParseOrThrow(byte[] source, Class<T> as) {
    return unpackOrThrow(parseOrThrow(source), as);
  }
}

File: csharp/src/Apache.Arrow.Flight.Sql/FlightSqlUtils.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Buffers;
using System.Collections.Generic;
using Arrow.Flight.Protocol.Sql;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;

namespace Apache.Arrow.Flight.Sql
{
    /// <summary>
    /// Helper methods for doing common Flight Sql tasks and conversions
    /// </summary>
    public class FlightSqlUtils
    {
        public static readonly FlightActionType FlightSqlCreatePreparedStatement = new("CreatePreparedStatement",
            "Creates a reusable prepared statement resource on the server. \n" +
            "Request Message: ActionCreatePreparedStatementRequest\n" +
            "Response Message: ActionCreatePreparedStatementResult");

        public static readonly FlightActionType FlightSqlClosePreparedStatement = new("ClosePreparedStatement",
            "Closes a reusable prepared statement resource on the server. \n" +
            "Request Message: ActionClosePreparedStatementRequest\n" +
            "Response Message: N/A");

        /// <summary>
        /// List of possible actions
        /// </summary>
        public static readonly List<FlightActionType> FlightSqlActions = new()
        {
            FlightSqlCreatePreparedStatement,
            FlightSqlClosePreparedStatement
        };

        /// <summary>
        /// Helper to parse {@link com.google.protobuf.Any} objects to the specific protobuf object.
        /// </summary>
        /// <param name="source">the raw bytes source value.</param>
        /// <returns>the materialized protobuf object.</returns>
        public static Any Parse(ByteString source)
        {
            return Any.Parser.ParseFrom(source);
        }

        /// <summary>
        /// Helper to unpack {@link com.google.protobuf.Any} objects to the specific protobuf object.
        /// </summary>
        /// <param name="source">the parsed Source value.</param>
        /// <typeparam name="T">IMessage</typeparam>
        /// <returns>the materialized protobuf object.</returns>
        public static T Unpack<T>(Any source) where T : IMessage, new()
        {
            return source.Unpack<T>();
        }

        /// <summary>
        /// Helper to parse and unpack {@link com.google.protobuf.Any} objects to the specific protobuf object.
        /// </summary>
        /// <param name="source">the raw bytes source value.</param>
        /// <typeparam name="T">IMessage</typeparam>
        /// <returns>the materialized protobuf object.</returns>
        public static T ParseAndUnpack<T>(ByteString source) where T : IMessage, new()
        {
            return Unpack<T>(Parse(source));
        }
    }

    /// <summary>
    /// A set of helper functions for converting encoded commands to IMessage types
    /// </summary>
    public static class FlightSqlExtensions
    {
        private static Any ParsedCommand(this FlightDescriptor descriptor)
        {
            return FlightSqlUtils.Parse(descriptor.Command);
        }

        private static IMessage UnpackMessage(this Any command)
        {
            if (command.Is(CommandStatementQuery.Descriptor))
                return FlightSqlUtils.Unpack<CommandStatementQuery>(command);
            if (command.Is(CommandPreparedStatementQuery.Descriptor))
                return FlightSqlUtils.Unpack<CommandPreparedStatementQuery>(command);
            if (command.Is(CommandGetCatalogs.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetCatalogs>(command);
            if (command.Is(CommandGetDbSchemas.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetDbSchemas>(command);
            if (command.Is(CommandGetTables.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetTables>(command);
            if (command.Is(CommandGetTableTypes.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetTableTypes>(command);
            if (command.Is(CommandGetSqlInfo.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetSqlInfo>(command);
            if (command.Is(CommandGetPrimaryKeys.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetPrimaryKeys>(command);
            if (command.Is(CommandGetExportedKeys.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetExportedKeys>(command);
            if (command.Is(CommandGetImportedKeys.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetImportedKeys>(command);
            if (command.Is(CommandGetCrossReference.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetCrossReference>(command);
            if (command.Is(CommandGetXdbcTypeInfo.Descriptor))
                return FlightSqlUtils.Unpack<CommandGetXdbcTypeInfo>(command);
            if (command.Is(TicketStatementQuery.Descriptor))
                return FlightSqlUtils.Unpack<TicketStatementQuery>(command);
            if (command.Is(TicketStatementQuery.Descriptor))
                return FlightSqlUtils.Unpack<TicketStatementQuery>(command);
            if (command.Is(CommandStatementUpdate.Descriptor))
                return FlightSqlUtils.Unpack<CommandStatementUpdate>(command);
            if (command.Is(CommandPreparedStatementUpdate.Descriptor))
                return FlightSqlUtils.Unpack<CommandPreparedStatementUpdate>(command);
            if (command.Is(CommandPreparedStatementQuery.Descriptor))
                return FlightSqlUtils.Unpack<CommandPreparedStatementQuery>(command);

            throw new ArgumentException("The defined request is invalid.");
        }

        /// <summary>
        /// Extracts a command from a FlightDescriptor
        /// </summary>
        /// <param name="descriptor"></param>
        /// <returns>An IMessage that has been parsed and unpacked</returns>
        public static IMessage? ParsedAndUnpackedMessage(this FlightDescriptor descriptor)
        {
            try
            {
                return descriptor.ParsedCommand().UnpackMessage();
            }
            catch (ArgumentException)
            {
                return null;
            }
        }

        public static ByteString Serialize(this IBufferMessage message)
        {
            int size = message.CalculateSize();
            var writer = new ArrayBufferWriter<byte>(size);
            message.WriteTo(writer);
            var schemaBytes = writer.WrittenSpan;
            return ByteString.CopyFrom(schemaBytes);
        }
    }
}
